sos/plugins/vdsm.py

```diff
@@ -0,0 +1,153 @@
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+
+import ast
+import os
+import stat
+import re
+
+from sos.plugins import Plugin, RedHatPlugin
+
+
+# This configuration is based on vdsm.storage.lvm.LVM_CONF_TEMPLATE.
+#
+# locking_type is set to 0 in order to match lvm sos commands. With this
+# configuration we don't take any locks, so we will never block because
+# there is a stuck lvm command.
+# locking_type=0
+#
+# use_lvmetad is set to 0 in order not to show cached, old lvm metadata.
+# use_lvmetad=0
+#
+# preferred_names and filter config values are set to capture Vdsm devices.
+# preferred_names=[ '^/dev/mapper/' ]
+# filter=[ 'a|^/dev/mapper/.*|', 'r|.*|' ]
+LVM_CONFIG = """
+global {
+    locking_type=0
+    use_lvmetad=0
+}
+devices {
+    preferred_names=[ "^/dev/mapper/" ]
+    ignore_suspended_devices=1
+    write_cache_state=0
+    disable_after_error_count=3
+    filter=[ "a|^/dev/mapper/.*|", "r|.*|" ]
+}
+"""
+LVM_CONFIG = re.sub(r"\s+", " ", LVM_CONFIG).strip()
+
+
+class Vdsm(Plugin, RedHatPlugin):
+    """vdsm Plugin"""
+
+    packages = (
+        'vdsm',
+        'vdsm-client',
+    )
+
+    plugin_name = 'vdsm'
+
+    def setup(self):
+        os.environ["LVM_SYSTEM_DIR"] = "/var/run/vdsm/lvm"
+        self.add_forbidden_path('/etc/pki/vdsm/keys/*')
+        self.add_forbidden_path('/etc/pki/vdsm/libvirt-spice/*-key.*')
+        self.add_forbidden_path('/etc/pki/libvirt/private/*')
+
+        self.add_cmd_output('service vdsmd status')
+
+        self.add_copy_spec(['/tmp/vds_installer*',
+                            '/tmp/vds_bootstrap*',
+                            '/etc/vdsm/*'])
+
+        limit = self.get_option('log_size')
+
+        if limit is not None:
+            self.add_copy_spec('/var/log/vdsm/*', sizelimit=limit)
+        else:
+            self.add_copy_spec('/var/log/vdsm/*')
+
+        self._add_vdsm_run_dir()
+        self.add_copy_spec(['/usr/libexec/vdsm/hooks',
+                            '/var/lib/vdsm',
+                            '/var/log/ovirt.log',
+                            '/var/log/sanlock.log'])
+
+        res = self.call_ext_prog('/usr/bin/pgrep qemu-kvm')
+        if res['status'] == 0:
+            for line in res['output']:
+                pid = line.strip()
+                self.add_copy_spec(["/proc/%s/cmdline" % pid,
+                                    "/proc/%s/status" % pid,
+                                    "/proc/%s/mountstats" % pid])
+        self.add_cmd_output(["/bin/ls -ldZ /etc/vdsm",
+                            "/bin/su vdsm -s /bin/sh -c\
+                             '/usr/bin/tree -l /rhev/data-center'",
+                             "/bin/su vdsm -s /bin/sh -c\
+                              '/bin/ls -lR /rhev/data-center'"])
```

Comment:
Why do any of these commands need to use a shell?

Reply:
also, please do not hard code paths in `sos` plugins unless it is absolutely necessary (i.e. a binary that is *never* expected to be in PATH, or in a standard `bin/` directory). We use a policy mechanism to control `PATH` to account for distribution variations and hard coding these defeats that.

Reply:
The commands need to be executed as user vdsm.

Reply:
Well, sure, but why do they need an absolute path? They should run in the `PATH` context of the `su`'d user.

Reply:
And why do they need a shell? That has nothing to do with switching user (unless you are relying on profile and rc expansion to establish an environment).

Reply:
Like I said, it is defined by the policy: you'll find that in the `sos.policy` module, specifically for Red Hat distributions in the file `sos/policies/redhat.py`. There are separate policies for RHEL, Fedora and other variants. Each of these may define its own path - this mechanism was introduced for two reasons:

  * to support UsrMove in Fedora and later RHEL
  * to isolate `sosreport` from broken `PATH` settings in the user environment

The general `RedHatPolicy` (parent class of the concrete product/distro classes) defines it like this:

```python
        # handle PATH for UsrMove
        if 'filesystem' not in pkgs:
            print("Could not find 'filesystem' package: "
                  "assuming PATH settings")
            usrmove = True
        else:
            filesys_version = pkgs['filesystem']['version']
            usrmove = True if filesys_version[0] == '3' else False

        if usrmove:
            self.PATH = "/usr/sbin:/usr/bin:/root/bin"
        else:
            self.PATH = "/sbin:/bin:/usr/sbin:/usr/bin:/root/bin"
        self.PATH += os.pathsep + "/usr/local/bin"
        self.PATH += os.pathsep + "/usr/local/sbin"
        self.set_exec_path()
```

You should find that this closely matches the ordering defined by the system. In my case, on Fedora, my system defined `PATH` includes `/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin`.

> when I run it 'sh' command on my computer it tries to run /usr/bin/sh instead of /bin/sh. I haven't made changes to PATH variable.

I don't know what your computer is running so I don't know what's correct, but this behaviour is consistent with recent Fedora and RHEL: `/bin` is just a symlink anyway, so what's the point of searching it?

```
$ ls -ld /bin
lrwxrwxrwx. 1 root root 7 Feb 10  2017 /bin -> usr/bin
```

Currently `sos` does not handle loading paths dynamically from `/etc/profile.d` but it is something we may add in future based on need (it's ugly since the only way to do it correctly is to invoke a shell and then ask it how `PATH` is defined - there's no way we will attempt to parse the shell drop-ins ourselves).

Reply:
OK, I guess the machine I used for testing wasn't configured well, I removed all the paths.